File: AtCoder/ABC/ABC426/G_Range_Knapsack_Query.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
#define endl '\n'

struct Item {
    int w, v;
    Item() : w(0), v(0) {}
    Item(int w, int v) : w(w), v(v) {}
};

struct Query {
    int qid, l, r, c;
    Query(int qid, int l, int r, int c) : qid(qid), l(l), r(r), c(c) {}
};

void solve() {
    int n, q, l, r, c;
    cin >> n;
    vector<Item> items(n + 1);
    for (int i = 1; i <= n; i++) cin >> items[i].w >> items[i].v;
    cin >> q;
    vector<LL> ans(q, -1);
    vector<Query> queries;
    for (int qid = 0; qid < q; qid++) {
        cin >> l >> r >> c;
        if (l == r) ans[qid] = (items[l].w <= c ? items[l].v : 0);
        else queries.push_back(Query(qid, l, r, c));
    }

    int MAX_C = 0;
    for (auto &query : queries) MAX_C = max(MAX_C, query.c);
    vector<vector<LL>> fl(n + 1, vector<LL>(MAX_C + 1, 0)), fr(n + 1, vector<LL>(MAX_C + 1, 0));
    // Divide and conquer
    auto dfs = [&](auto &&dfs, int left, int right, vector<Query> &queries) -> void {
        if (queries.empty() || left >= right) return;
        int mid = left + ((right - left) >> 1);
        // Divide into three parts: left, across middle, right
        vector<Query> ql, qm, qr;
        for (auto &query : queries) {
            if (query.l <= mid && query.r > mid) qm.push_back(query);
            else if (query.r <= mid) ql.push_back(query);
            else qr.push_back(query);
        }
        dfs(dfs, left, mid, ql);
        dfs(dfs, mid + 1, right, qr);
        if (!qm.empty()) {
            int C = ranges::max(qm | views::transform(&Query::c));
            for (int c = 0; c <= C; c++)
                fl[mid + 1][c] = fr[mid][c] = 0;
            // fl: knapsack on [left, mid] (from right to left)
            for (int i = mid; i >= left; i--) {
                auto [w, v] = items[i];
                for (int c = 0; c <= C; c++) {
                    if (c >= w) fl[i][c] = max(fl[i + 1][c], fl[i + 1][c - w] + v);
                    else fl[i][c] = fl[i + 1][c];
                }
            }
            // fr: knapsack on [mid + 1, right] (from left to right)
            for (int i = mid + 1; i <= right; i++) {
                auto [w, v] = items[i];
                for (int c = 0; c <= C; c++) {
                    if (c >= w) fr[i][c] = max(fr[i - 1][c], fr[i - 1][c - w] + v);
                    else fr[i][c] = fr[i - 1][c];
                }
            }
            // Merge the results of fl and fr for each query
            for (auto &[qid, l, r, c] : qm)
                for (int cc = 0; cc <= c; cc++)
                    ans[qid] = max(ans[qid], fl[l][cc] + fr[r][c - cc]);
        }
        return;
    };
    dfs(dfs, 1, n, queries);
    for (auto &x : ans) cout << x << endl;
    return;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    solve();
    return 0;
}

File: AtCoder/ABC/ABC426/G_Range_Knapsack_Query.py
"""
Divide and Conquer + Knapsack DP
Source: https://www.luogu.com.cn/problem/P6240 (貓樹)

Knapsack DP 有容易合併的性質，所以可以離線 + 分治來做。
- 對於穿過中間點 mid 的詢問，可以計算從中點往左和往右的 DP 結果，然後合併。
- 對於詢問區間完全在左邊或右邊的，往下分治處理。

每個物品會出現在 O(log n) 個分治區間內，每次出現計算背包時會貢獻 O(C) 的時間複雜度，總共有 N 個物品。
每次詢問需要花 O(C) 的時間複雜度，總共有 Q 個詢問。
所以總時間複雜度為 O(N * C * log N + Q * C)。

因為 1e5 * 1e3 = 1e8，對 Python 來說有點勉強，需要考慮一些常數優化。
"""
from typing import List

N = int(input())
items = [(0, 0)] + [tuple(map(int, input().split())) for _ in range(N)]
Q = int(input())
queries = [tuple(map(int, input().split())) for _ in range(Q)]

ans = [-1] * Q
MAX_C = 0
qids = []
for qid, (l, r, c) in enumerate(queries):
    if l == r:
        ans[qid] = items[l][1] if items[l][0] <= c else 0
    else:
        qids.append(qid)
        MAX_C = max(MAX_C, c)

fl = [[0] * (MAX_C + 1) for _ in range(N + 1)]
fr = [[0] * (MAX_C + 1) for _ in range(N + 1)]

def solve(left: int, right: int, qids: List[int]) -> None:
    if len(qids) == 0:
        return
    mid = (left + right) // 2
    ql, qr, qm = [], [], []
    C = 0
    for qid in qids:
        l, r, c = queries[qid]
        if l <= mid and r > mid:
            qm.append(qid)
            C = max(C, c)
        elif r <= mid:
            ql.append(qid)
        else:
            qr.append(qid)
    if len(qm) > 0:
        for c in range(MAX_C + 1):
            fl[mid + 1][c] = fr[mid][c] = 0
        # fl: [left, mid]
        for i in range(mid, left - 1, -1):
            w, v = items[i]
            prev, curr = fl[i + 1], fl[i]
            for c in range(C + 1):
                curr[c] = prev[c]
                if c >= w:
                    curr[c] = max(prev[c], prev[c - w] + v)
        # fr: [mid + 1, right]
        for i in range(mid + 1, right + 1):
            w, v = items[i]
            prev, curr = fr[i - 1], fr[i]
            for c in range(C + 1):
                curr[c] = prev[c]
                if c >= w:
                    curr[c] = max(curr[c], prev[c - w] + v)
        for qid in qm:
            l, r, c = queries[qid]
            mx = 0
            fll, frr = fl[l], fr[r]
            for cc in range(c + 1):
                mx = max(mx, fll[cc] + frr[c - cc])
            ans[qid] = mx
    solve(left, mid, ql)
    solve(mid + 1, right, qr)

solve(1, N, qids)
print(*ans, sep='\n')
